### PR TITLE
Allow using salt modules in onlyif and unless

### PR DESCRIPTION
**Pull request by gtmanfred**
### What does this PR do?
Allow using salt modules in onlyif and unless

### What issues does this PR fix or reference?
Implements #5050 
Implements #51604
### Previous Behavior
Can only use commands that run through cmd.retcode

### New Behavior
Can use any salt module, and the result is determined by the trueish or falsish return of the module.

### Tests written?

Yes

### Commits signed with GPG?

Yes

### CODE REVIEW

**Comment by gtmanfred** [comment]
@saltstack/team-core can I get a review on this.

Thanks!
Daniel

File: salt/state.py [hunk 1]
```diff
@@ -883,19 +883,31 @@ def _run_check_onlyif(self, low_data, cmd_opts):
...
+                                'skip_watch': True,
+                                'result': True})
+                elif cmd == 0:
+                    ret.update({'comment': 'onlyif condition is true', 'result': False})
+            elif isinstance(entry, dict):
+                if 'fun' not in entry:
+                    ret['comment'] = 'no `fun` argument in onlyif: {0}'.format(entry)
+                    log.warning(ret['comment'])
+                    return ret
+                result = self.functions[entry.pop('fun')](**entry)
```
Comment (mattp-) [inline salt/state.py]: nice 👍 i'd suggest adding a check for __context__['retcode'] and taking that over the return result if set

Comment (gtmanfred) [reply salt/state.py]: ahh good point.

Comment (gtmanfred) [reply salt/state.py]: going to make it only care about if retcode is anything but 0.

Comment (gtmanfred) [reply salt/state.py]: since otherwise, we would always be 0, because the salt minion sets it to 0 if nothing is set.

Comment (gtmanfred) [reply salt/state.py]: Done!

**Comment by gtmanfred** [comment]
Ok, this should be good to merge 😄

**Review by dwoz — APPROVED** [review]

**Comment by mchugh19** [comment]
This is pretty major and should get a shout out in the release notes too

**Comment by max-arnold** [comment]
@gtmanfred Unfortunately, this doesn't work consistently across all states and needs to be improved. The following states have their own `mod_run_check` implementations:

* `states.git`
* `states.cmd`
* `states.macpackage`
* `states.file`
* `states.docker_container`

This is where unless/onlyif get skipped: https://github.com/saltstack/salt/blob/develop/salt/state.py#L1955-L1957

I discovered this while trying to fix slots #53307

**Comment by mchugh19** [comment]
> The following states have their own `mod_run_check` implementations

In glancing over those states, it's not immediately obvious why they have their own `mod_run_check` implementations? Would functionality change if they were just removed?

**Comment by gtmanfred** [comment]
They had unless and onlyif implementations before unless and onlyif was implemented, and unless/onlyif was a simpler implementation, so instead of having to make compatible changes for everything that the cmd.run onlyif did, i just let it be configured to be able to customize it.

**Pull request merged**